buildkonfig-gradle-plugin/src/test/kotlin/com/codingfeline/buildkonfig/gradle/BuildKonfigPluginConstFieldsTest.kt

```diff
@@ -0,0 +1,101 @@
+package com.codingfeline.buildkonfig.gradle
+
+import com.google.common.truth.Truth
+import org.gradle.testkit.runner.GradleRunner
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import java.io.File
+
+class BuildKonfigPluginConstFieldsTest {
+
+    @get:Rule
+    val projectDir = TemporaryFolder()
+
+    lateinit var buildFile: File
+
+    lateinit var settingFile: File
+
+    private val buildFileHeader = """
+        |plugins {
+        |    id("kotlin-multiplatform")
+        |    id("com.codingfeline.buildkonfig")
+        |}
+        |
+        |repositories {
+        |   mavenCentral()
+        |}
+        |
+    """.trimMargin()
+
+    private val buildFileMPPConfig = """
+        |kotlin {
+        |  jvm()
+        |  js(IR) {
+        |    browser()
+        |    nodejs()
+        |  }
+        |}
+    """.trimMargin()
+
+    @Before
+    fun setup() {
+        buildFile = projectDir.newFile("build.gradle.kts")
+        settingFile = projectDir.newFile("settings.gradle")
+        settingFile.writeText(settingsGradle)
+    }
+
+    @Test
+    fun `const values for expect object contains suppress annotation`() {
+        buildFile.writeText(
+            """
+            |import com.codingfeline.buildkonfig.compiler.FieldSpec.Type
+            |$buildFileHeader
+            |
+            |buildkonfig {
+            |   packageName = "com.example"
+            |   
+            |   defaultConfigs {
+            |       buildConfigConstField(Type.STRING, "foo", "defaultValue")
```

Comment:
Can we instead have a global flag to enable const value support for all fields, as I think people basically want every field to be a const?

```groovy
buildkonfig {
  packageName = "com.example"
  experimentalConstValue = true

  defaultConfigs {
    buildConfigField(Type.STRING, "foo", "defaultValue")
  }
}
```

this way we can make sure that this support is experimental and that migration is easy.

Reply:
I believe there is usecases when fields shouldn't be const.

Let's take for example library version information that comes from buildkonfig.
If it will be a const there is no way to read library version in runtime, as const variable will be inlined in call-site.
It can be workarounded with defining some non-const val or fun at library side and using it instead of const val though.

Maybe there are some more usecases rhat requires non-const vals?...

Reply:
To be honest I'll like to suggest some dsl imporvements in field declaration later, to make it more like gradle default api, for example:
Groovy DSL
```groovy
buildConfigField type: Type.String, name: "foo", value: "bar"
buildConfigField Type.String, "foo", "bar", const: true
buildConfigField type: Type.String, name: "foo", value: "bar", nullable: true
```
Kotlin DSL
```groovy
buildConfigField(type = Type.String, name = "foo", value = "bar")
buildConfigField(Type.String, "foo", "bar", const = true)
buildConfigField(type = Type.String, name = "foo", value = "bar", nullable = true)
```

Reply:
> Let's take for example library version information that comes from buildkonfig.

[The official Kotlin document](https://kotlinlang.org/docs/properties.html#compile-time-constants) says

> The compiler will inline usages of the constant, replacing the reference to the constant with its actual value. However, the field will not be removed and therefore can be interacted with using reflection.

This means we still have actual fields and can access it, so all variables can safely have a const modifier, right?

Or am I missing something?


---

> To be honest I'll like to suggest some dsl improvements in field declaration later, to make it more like gradle default API

this looks nice 👍

Reply:
> This means we still have actual fields and can access it, so all variables can safely have a const modifier, right?
> Or am I missing something?

Let's imagine we have a library `foo` with `const val VERSION = "1.0.0"` and some consumer of a library (another library) named `bar` that depends on `foo` and uses that `VERSION` field somewhere. After compiling `bar` will contains inlined `VERSION` field (as intended). But after compiling some third-party target (app, who depends on both `foo` and `bar`) wants to upgrade `foo` dependency to `1.1.0`, it will ship with fresh `VERSION = "1.1.0"` field, but `bar` doesn't have any way to check actual `foo` version at runtime (as any calls to `VERSION` is inlined at compile-time). So, although new version contains new vals, consumers of previous versions of `foo` have no any way to get *real* const values, unless `foo` have some non-const wrapper specially for this field (like `val BuildKonfig.RUNTIME_VERSION get() = VERSION`).

Reply:
So, I'm totally okay with global const flag and opt-out behavior per-field, it will be pretty easy with new DSL I suggested. The only question is that DSL should be a part of this PR or separate one?

Reply:
Ok, convinced 👍 

Let's do DSL refinement in another PR.